src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -13,12 +14,47 @@ namespace Aspire.Hosting.ApplicationModel;
 public class ResourceNotificationService(ILogger<ResourceNotificationService> logger)
 {
+    private static readonly TimeSpan s_defaultWaitForResourceStateTimeout = TimeSpan.FromSeconds(60);
+    // For each resource we have a TCS for each state we see it move to or a caller is waiting for it to move to
+    private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, TaskCompletionSource>> _resourcesStateChanges = new(StringComparers.ResourceName);
     // Resource state is keyed by the resource and the unique name of the resource. This could be the name of the resource, or a replica ID.
     private readonly ConcurrentDictionary<(IResource, string), ResourceNotificationState> _resourceNotificationStates = new();
     private readonly ILogger<ResourceNotificationService> _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+    private readonly CancellationToken _applicationStopping;
 
     private Action<ResourceEvent>? OnResourceUpdated { get; set; }
 
+    /// <summary>
+    /// Creates a new instance of <see cref="ResourceNotificationService"/>.
+    /// </summary>
+    /// <param name="logger">The logger.</param>
+    /// <param name="hostApplicationLifetime">The host application lifetime.</param>
+    public ResourceNotificationService(ILogger<ResourceNotificationService> logger, IHostApplicationLifetime hostApplicationLifetime)
+        : this(logger)
+    {
+        ArgumentNullException.ThrowIfNull(hostApplicationLifetime);
+
+        _applicationStopping = hostApplicationLifetime.ApplicationStopping;
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach the specified state. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches the specified state. If the resource is already in the target state, the method will return immediately.<br/>
+    /// If the resource never reaches the target state, the method will wait for the passed <see cref="CancellationToken"/> to be cancelled, or 60 seconds if <c>null</c>, and then throw <see cref="TaskCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resouce.</param>
+    /// <param name="targetState">The state to wait for the resource to transition to. See <see cref="KnownResourceStates"/> for common states. Defaults to <see cref="KnownResourceStates.Running"/>.</param>
+    /// <param name="cancellationToken">The cancellation token to monitor to cancel the wait operation. If <c>null</c> defaults to waiting for 60 seconds.</param>
+    /// <returns></returns>
+    public Task WaitForResourceAsync(string resourceName, string? targetState = null, CancellationToken? cancellationToken = null)
+    {
+        var stateToWaitFor = targetState ?? KnownResourceStates.Running;
+        var token = cancellationToken ?? new CancellationTokenSource(s_defaultWaitForResourceStateTimeout).Token;
```

Comment:
you should dispose this CTS if you create it

Reply:
Updated.